### PR TITLE
Compress compat version in compat logs

### PR DESCRIPTION
**Pull request by oxinabox**
This does one of the suggestions from https://github.com/JuliaLang/Pkg.jl/issues/1855

I think it is a big improvement.

- [x] It needs tests.
- [x] Especially around the edge cases where it is broken into multiple discontigious runs

### Current Master
```julia
julia> using Pkg; pkg"add DataFrames@0.17 CSV@0.5"
#...
ERROR: Unsatisfiable requirements detected for package CSV [336ed68f]:
 CSV [336ed68f] log:
 ├─possible versions are: [0.3.0-0.3.1, 0.4.0-0.4.3, 0.5.0-0.5.26, 0.6.0-0.6.2] or uninstalled
 ├─restricted to versions 0.5 by an explicit requirement, leaving only versions 0.5.0-0.5.26
 └─restricted by compatibility requirements with DataFrames [a93c6f00] to versions: [0.3.0-0.3.1, 0.4.0-0.4.3] or uninstalled — no versions left
   └─DataFrames [a93c6f00] log:
     ├─possible versions are: [0.11.7, 0.12.0, 0.13.0-0.13.1, 0.14.0-0.14.1, 0.15.0-0.15.2, 0.16.0, 0.17.0-0.17.1, 0.18.0-0.18.4, 0.19.0-0.19.4, 0.20.0-0.20.2, 0.21.0-0.21.2] or uninstalled
     └─restricted to versions 0.17 by an explicit requirement, leaving only versions 0.17.0-0.17.1
```

### With this PR
```julia
julia> using Pkg; pkg"add DataFrames@0.17 CSV@0.5"
#...
ERROR: Unsatisfiable requirements detected for package CSV [336ed68f]:
 CSV [336ed68f] log:
 ├─possible versions are: 0.3.0-0.6.2 or uninstalled
 ├─restricted to versions 0.5 by an explicit requirement, leaving only versions 0.5.0-0.5.26
 └─restricted by compatibility requirements with DataFrames [a93c6f00] to versions: 0.3.0-0.4.3 or uninstalled — no versions left
   └─DataFrames [a93c6f00] log:
     ├─possible versions are: 0.11.7-0.21.2 or uninstalled
     └─restricted to versions 0.17 by an explicit requirement, leaving only versions 0.17.0-0.17.1
```

### CODE REVIEW

**Comment by KristofferC** [comment]
Hm, wiring `possible versions are: 0.3.0-0.6.2` is pretty pointless to write since it is quite rare just the first and last version is of any particular use. The previous printing was compactified but still gave you all information of available versions.

**Comment by oxinabox** [comment]
> Hm, wiring possible versions are: 0.3.0-0.6.2 is pretty pointless to write since it is quite rare just the first and last version is of any particular use. The previous printing was compactified but still gave you all information of available versions.

Our experience of resolving these seems to be very different.
I only ever need to know the range support, so i know how much i need to adjust version bounds.
What am I going to do with a list of all available versions?
I personally have never had a use for that list when I am resolving this.

I can find a list of all avaible version by looking at release notes on github (and intersecting with the range shown), which will also have release notes and will often tell me more about why compat was dropped etc.

**Comment by oxinabox** [comment]
Or are you specifically talking about the "All possible versions" line?
rather than the other lines about restrictions?

I find that line almost entirely useless and was intending a follow up PR to discuss deleting it. 
It is on the list in #1855.
It does have some use to contextualize some of the other lines about restrictions.
So I am not sure deleting it is nessecarily the best thing to do, but compressing it seems like a net win to decrase how much noise it is adding

**Comment by singularitti** [comment]
What will this PR print if you have a minor version skipped, e.g., `[0.6, 0.7, 0.8, 1.0]` and `[0.4, 0.6, 0.8]`? Will this lead users try to add `0.6, 0.7, 0.8, 0.9, 1.0` and `0.4, 0.5, 0.6, 0.7, 0.8` to their `[compat]` fields in `Project.toml`?

Could a compromised solution be like:
```julia
julia> using Pkg; pkg"add DataFrames@0.17 CSV@0.5"
#...
ERROR: Unsatisfiable requirements detected for package CSV [336ed68f]:
 CSV [336ed68f] log:
 ├─possible versions are: [0.3+, 0.4+, 0.5+, 0.6+] or uninstalled
 ├─restricted to versions 0.5 by an explicit requirement, leaving only versions 0.5.0-0.5.26
 └─restricted by compatibility requirements with DataFrames [a93c6f00] to versions: [0.3+, 0.4+] or uninstalled — no versions left
   └─DataFrames [a93c6f00] log:
     ├─possible versions are: [0.11.7, 0.12+, 0.13+, 0.14+, 0.15+, 0.16+, 0.17+, 0.18+, 0.19+, 0.20+, 0.21+] or uninstalled
     └─restricted to versions 0.17 by an explicit requirement, leaving only versions 0.17.0-0.17.1
```
What usually matters to me when writing `[compat]` is the major and minor versions, I don't care about patch versions.

**Comment by DilumAluthge** [comment]
Personally I (like @KristofferC) find it helpful to have all of the versions listed out. I don't particular like this "compressed range" format- it hides too much information.

Just my personal preference.

**Comment by DilumAluthge** [comment]
I think that e.g. adding colors and stuff like that will help make these messages more readable.

But personally I don't want to remove any information from the messages.

**Comment by oxinabox** [comment]
> What will this PR print if you have a minor version skipped, e.g., `[0.6, 0.7, 0.8, 1.0]` or `[0.4, 0.6, 0.8]`? Will this lead users try to add `0.6, 0.7, 0.8, 0.9, 1.0` and `0.4, 0.5, 0.6, 0.7, 0.8` to their `[compat]` fields in `Project.toml`? I have seen packages skipping versions.


It is compressed relative to what exists so question can't be answered without that.
From the tests: if everything that exists is:
`[v"1.0.0", v"1.1.0", v"1.2.0", v"1.2.1", v"2.0.0", v"2.0.1", v"3.0.0", v"3.1.0"]`

and it is compatible with 1.x and 3.x but not 2.x, i.e. with:
`[v"1.0.0", v"1.1.0", v"1.2.0", v"1.2.1", v"2.0.0", v"2.0.1", v"3.0.0", v"3.1.0"]`

Then the shown versions will be: `1.0.0-1.2.1, 3.0.0-3.1.0`


---
> ...  I don't want to remove any information from the messages.

We need to balance the amount of information.
There is a trade-off between presenting every piece of information we know, vs keeping the amount of information presented from being overwelming.
We want to optimize the signal to noise ratio, presenting the useful facts that change what actions people do (to more efficent ones), without presenting too much excess that they ignore

@DilumAluthge  for my own understanding can you explain how you use the version numbers in resolving conflicts?
I think understanding your process  (which is clearly different from mine) will help understand what informat is signal, and what is noise.


My process is:
1. if it is a direct dependency, look at possible versions  and versions allowed under restrictions, and if I am not getting to the top then see about relaxing those requirements.
2. if it is further down stream, then use the bounds to workout what is being most restrictive and go to github for that package where I am going to look at Project.tomls, potentially on various tagged releases, so I can see what is allowed. I am potentially looking forward and back and at master to see what is allowed when. (This is also why it is so annoying to me to have the Project.toml and registry out of sync). From the Github page I am likely to then make a PR.

So I have no need to know about every single release in the Pkg screen, because I am about to go to github where i can look at those relases and immidately see even more information about them -- each releases exact compatibilities.

**Comment by DilumAluthge** [comment]
> 2\. if it is further down stream, then use the bounds to workout what is being most restrictive and go to github for that package where I am going to look at Project.tomls, potentially on various tagged releases, so I can see what is allowed. I am potentially looking forward and back and at master to see what is allowed when. (This is also why it is so annoying to me to have the Project.toml and registry out of sync). From the Github page I am likely to then make a PR.
> 
> So I have no need to know about every single release in the Pkg screen, because I am about to go to github where i can look at those relases and immidately see even more information about them -- each releases exact compatibilities.

You can't do that. You have to go to the General registry, go to each package, open up Versions.toml and Compat.toml, etc.

The registry is the ultimate source of truth. You cannot rely on Git tags/GitHub releases, and you cannot rely on Project.toml files. The Pkg resolver uses the General registry only, so that's what I have to look at.

So if we don't print that information in the error message, it forces me to go to the General registry. It's much better in my opinion to just have these versions listed out in the error message.

@KristofferC What do you think?

**Comment by oxinabox** [comment]
@dillumaluthge
Could you tell me your process?
Including, but not only what you use these numbers for

File: src/versions.jl [hunk 1]
```diff
@@ -273,6 +275,42 @@ end
...
+"""
+    compressed_versionspec(pool, subset=pool)
+
+ - `pool`: all versions that exist (a superset of `subset`)
+ - `subset`: the subset of those versions that we want to include in the spec.
+
+Works out a compressed `VersionSpec` that permits everything in the `subset`,
+and nothing else from the `pool`.
+"""
+function compressed_versionspec(pool, subset=pool)
```
Comment (oxinabox) [inline src/versions.jl:287]: Pkg-Dev call wants to rename this because it is not quite the same as what compress means in the registry tools.
Maybe: `ranged_versionspec` ?

**Comment by StefanKarpinski** [comment]
I'm ok with this since although the "possible versions" have less information, the relevant information is still included on the following "restricted to" and "restricted by" lines. Using the example in the PR body, yes, `0.3.0-0.6.2` as a list of possible versions is a bit vague, but then you immediately see what the important subsets of that are, namely `0.5.0-0.5.26` and `0.3.0-0.4.3`. Does it matter whether a `0.3.2` release exists? No, it does not. That is not relevant information. What matters is that these two ranges are disjoint. If there were a `0.3.2` version and it was compatible, then, yes, that would be relevant, but then again (a) these ranges wouldn't be printed the way they are and (b) we wouldn't be seeing this resolver error in the first place. In short, I'm in favor of this change as it makes it easier to see what's happening with a resolver error.

File: src/versions.jl (hunk 1 above)
Comment (StefanKarpinski) [reply src/versions.jl:287]: `versionspec_range_string`?

**Comment by KristofferC** [comment]
Ok, I'll know which people to cc then for debugging / figuring out resolver errors.

**Comment by StefanKarpinski** [comment]
Stop being cranky 🙂

**Comment by oxinabox** [comment]
Ok should be good to go now.

**Comment by KristofferC** [comment]
Wow, I did not mean to close the PR... That was a finger slip. Sorry about that.

**Comment by StefanKarpinski** [comment]
Let's give it a try and see how it goes 🤷‍♂️. It's hard to predict if something like this will be more or less confusing to people without testing it out on actual people.
